File: csharp/OutsideInDemo/Implementation/Classic/FizzBuzz.cs
using System.Globalization;

namespace Implementation.Classic
{
    public class FizzBuzz
    {
        public string CalculateFor(int input)
        {
            if (input.IsDivisibleBy(3) && input.IsDivisibleBy(5))
            {
                return "Fizz-Buzz";
            }

            if (input.IsDivisibleBy(3))
            {
                return "Fizz";
            }

            if (input.IsDivisibleBy(5))
            {
                return "Buzz";
            }

            return input.ToString(CultureInfo.InvariantCulture);
        }
    }
}

File: kotlin/fizzbuzz-functional/src/main/kotlin/FizzBuzz.kt
import kotlin.coroutines.experimental.buildSequence

class Rule(val appliesTo: (Int) -> Boolean, val resultFor: (Int) -> String)

fun generate(input: Int): String = allRules()
        .filter { it.appliesTo(input) }
        .map { it.resultFor(input) }
        .first()

private fun allRules() = buildSequence {
    yield(isDivisibleBy(3, 5) to returns("Fizz-Buzz"))
    yield(isDivisibleBy(5) to returns("Buzz"))
    yield(isDivisibleBy(3) to returns("Fizz"))
    yield(otherwise to numberItself)
}.map { Rule(it.first, it.second) }

private val otherwise = { _: Int -> true }
private val numberItself = { input: Int -> input.toString() }
private val returns = { constantValue: String -> { _: Int -> constantValue } }

private fun isDivisibleBy(vararg denominators: Int) =
        { input: Int -> denominators.all { input % it == 0 } }
